### PR TITLE
Change type for max function in RateLimitOptions to support async

### LINKED ISSUE #176: Wrong type for max function in RateLimitOptions
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure it has not already been reported

### Fastify version

3.19.1

### Plugin version

5.5.0

### Node.js version

16.6.2

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

Big Sur 11.4

### Description

The docs page states that the max function `can be an async function with the signature async (req, key) => {} where req is the Fastify request object and key is the value generated by the keyGenerator. The function must return a number.`

However, typescript complains I have the wrong type when I use an async function because Promise<number> is not assignable to number. 

Basically, I think the type definition for the max function needs to be changed from:

```js
number | ((req: FastifyRequest, key: string) => number)
```

to:

```js
number | ((req: FastifyRequest, key: string) => number) | ((req: FastifyRequest, key: string) => Promise<number>)
```

### Steps to Reproduce

```js
// Declare a config object
const config: RateLimitPluginOptions = {
    ban: 10,
    timeWindow: '2 minute',
    store: new Redis(process.env.REDIS_URL, { connectTimeout: 500, maxRetriesPerRequest: 1 }) as any,
    keyGenerator: function (req) {
        return req.headers['authorization'] || req.ip;
    },

    // This is where the error is
    async max(request, key) {
        const apiKeyObject = await ApiKey.findOne({ where: { apiKey: key } });
        if (!apiKeyObject) {
            return 10;
        }

        return apiKeyObject.rateLimit;
    }
};

// Register the plugin
app.register(fastifyRateLimit, config);
```

The exact error is:
```
Type '(request: FastifyRequest<RouteGenericInterface, Server, IncomingMessage>, key: string) => Promise<number>' is not assignable to type '(req: FastifyRequest<RouteGenericInterface, Server, IncomingMessage>, key: string) => number'.
    Type 'Promise<number>' is not assignable to type 'number'.
```

### Expected Behavior

_No response_

### PR DESCRIPTION
**Pull request by leonitousconforti**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
Can you add a test to test/types/types.test.ts?

**Comment by leonitousconforti** [comment]
I've never written tests for type declaration files, so I'm not sure if this is how you would do it, but I gave it a shot

**Review by mcollina — APPROVED** [review]
lgtm

**Review by Eomm — APPROVED** [review]

File: test/types/types.test.ts [hunk 1]
```diff
@@ -45,7 +45,7 @@ const options2 = {
 
 const options3 = {
   global: true,
-  max: (req: FastifyRequest<RequestGenericInterface>, key: string) => (42),
+  max: (req: FastifyRequest<RequestGenericInterface>, key: string) => Promise.resolve(42),
```
Comment (Eomm) [inline test/types/types.test.ts:48]: Could you add a `option4` object: changing this one we are no more testing the number

**Review by Eomm — CHANGES_REQUESTED** [review]

File: test/types/types.test.ts (hunk 1 above)
Comment (leonitousconforti) [reply test/types/types.test.ts:48]: Is there a need for an options4, it seems like options 2 already tests the number to me
https://github.com/fastify/fastify-rate-limit/blob/ded16faae4ae4e71b499a108dbab9002e96818af/test/types/types.test.ts#L39-L44

Comment (leonitousconforti) [reply test/types/types.test.ts:55]: I did change it to what you asked, if there was a reason I'm just not seeing

**Review by Eomm — APPROVED** [review]

**Pull request merged**